storage/aptosdb/src/state_store/mod.rs

```diff
@@ -1177,10 +1194,47 @@ impl StateValueWriter<StateKey, StateValue> for StateStore {
     }
 
     fn get_progress(&self, version: Version) -> Result<Option<StateSnapshotProgress>> {
-        Ok(self
+        let main_db_progress = self
             .state_kv_db
             .metadata_db()
             .get::<DbMetadataSchema>(&DbMetadataKey::StateSnapshotRestoreProgress(version))?
-            .map(|v| v.expect_state_snapshot_progress()))
+            .map(|v| v.expect_state_snapshot_progress());
+
+        // verify if internal indexer db and main db are consistent before starting the restore
+        if self.internal_indexer_db.is_some()
+            && self
+                .internal_indexer_db
+                .as_ref()
+                .unwrap()
+                .statekeys_enabled()
+        {
+            let progress_opt = self
+                .internal_indexer_db
+                .as_ref()
+                .unwrap()
+                .get_restore_progress(version)?;
+
+            match (main_db_progress, progress_opt) {
+                (None, None) => (),
+                (Some(main_progres), Some(indexer_progress)) => {
+                    if main_progres.key_hash > indexer_progress.key_hash {
+                        bail!(
+                            "Inconsistent restore progress between main db and internal indexer db. main db: {:?}, internal indexer db: {:?}",
+                            main_progres,
+                            indexer_progress,
+                        );
+                    }
+                },
+                _ => {
```

Comment:
It's a valid case when the indexer has the first progress write done, while the main db didn't see the progress?

Reply:
yes, I assume the main db will eventually catch up.

Reply:
If the node died between the first indexer write and the first main db write, on restart, this code will fail and it will never recover.